datafusion/functions-aggregate-common/src/aggregate/array_agg.rs

```diff
@@ -0,0 +1,491 @@
+    ) -> Result<()> {
+        let singular_col = values
+            .first()
+            .ok_or(internal_datafusion_err!("invalid agg input"))?;
+
+        self.stacked_batches.push(Arc::clone(singular_col));
+        let batch_index = self.stacked_batches.len() - 1;
+
+        if let Some(filter) = opt_filter {
+            for (array_offset, (group_index, filter_value)) in
+                group_indices.iter().zip(filter.iter()).enumerate()
+            {
+                if let Some(true) = filter_value {
+                    self.stacked_group_indices.push((
+                        *group_index,
+                        batch_index,
+                        array_offset,
+                    ));
+                }
+            }
+        } else {
+            for (array_offset, group_index) in group_indices.iter().enumerate() {
+                self.stacked_group_indices.push((
+                    *group_index,
+                    batch_index,
+                    array_offset,
+                ));
+            }
+        }
+        self.indice_sorted = false;
+        self.max_group = self.max_group.max(total_num_groups);
+        Ok(())
+    }
+
+    fn evaluate(&mut self, emit_to: EmitTo) -> Result<ArrayRef> {
+        let arr = self.consume_stacked_batches(emit_to)?;
+        Ok(Arc::new(arr) as ArrayRef)
+    }
+
+    // filtered_null_mask(opt_filter, &values);
+    fn state(&mut self, emit_to: EmitTo) -> Result<Vec<ArrayRef>> {
+        Ok(vec![self.evaluate(emit_to)?])
+    }
+
+    fn merge_batch(
+        &mut self,
+        values: &[ArrayRef],
+        group_indices: &[usize],
+        // for merge_batch which happens at final stage
+        // opt_filter will always be none
+        _opt_filter: Option<&BooleanArray>,
+        total_num_groups: usize,
+    ) -> Result<()> {
+        let singular_col = values
+            .first()
+            .ok_or(internal_datafusion_err!("invalid agg input"))?;
+        let list_arr = singular_col.as_list::<i32>();
+        let new_array_number = self.stacked_batches.len();
+        // TODO: the backed_arr contains redundant data
+        // make sure that flatten_group_index has the same length with backed_arr
+        let flatten_group_index =
+            group_indices
+                .iter()
+                .enumerate()
+                .flat_map(|(row, group_index)| {
+                    let end = list_arr.value_offsets()[row + 1].as_usize();
+                    let start = list_arr.value_offsets()[row].as_usize();
+                    (start..end).map(|offset| (*group_index, new_array_number, offset))
+                });
+        self.stacked_group_indices.extend(flatten_group_index);
+
+        let backed_arr = list_arr.values();
+        self.stacked_batches.push(Arc::clone(backed_arr));
+        self.indice_sorted = false;
+        self.max_group = self.max_group.max(total_num_groups);
+        Ok(())
+    }
+
+    fn size(&self) -> usize {
+        size_of_val(self)
```

Comment:
> some error on memory reservation, i'm not sure if the calculation if `fn size` is wrong, but as i understand, we only account for the buffer newly created by the implementation, not the buffer we borrowed somewhere (i.e the stacked ArrayRef everytime we receive from `fn merge_batch` or `fn update_batch`. Maybe it's a good chance for me learn how mem reservation/spilling works

I took a quick look at this code -- One thing we probably need to account for is the memory in the held ArrayRefs -- specifically by calling https://docs.rs/arrow/latest/arrow/array/trait.Array.html#tymethod.get_array_memory_size on all the stacked arrays



However,  I bet the fuzz test failure is due to actually better accounting of memory. Maybe we need to readjust the parameters or something